046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,63 @@
+
+## Proposal
+
+The proposal is a REST API endpoint on the KafkaAgent which exposes the BrokerState metric that it already uses. And modify KafkaRoller so that it queries the new endpoint to find out what state a broker is in. 
+
+1. Add a web server to KafkaAgent that exposes an endpoint which returns a value of metric
+    - Configure Jetty server to only accept SSL connections on a port 8443.
+    - Expose an endpoint `/broker-state`.
+    - Add a servlet to handle a `GET` request and return an integer value of `BrokerState` metric.
+    - It will reuse broker's own certificate, which will already be in the container.
+2. Update KafkaRoller to query the broker state from the KafkaAgent API
+    - Currently KafkaRoller obtains headless service for a given pod and uses it to boostrap Kafka AdminClient. It can use the same mechanism to get the address for the KafkaAgent endpoint.
+    - Instead of checking broker's readiness through `isReady()` function, utilise the new restful endpoint to determine if the broker is ready.
+    - After restarting a broker if it does not become ready within the operational timeout period, check whether the broker is performing log recovery by querying the new endpoint:
+        - If log recovery is in progress, do not restart the broker.
+        - Otherwise, force restart the broker which is the current behabviour.
+    - Add a new configurable timeout for log recovery that is 60 minutes by default.
+    - If the log recovery does not complete within the timeout period, then fail the reconciliation and report an error for the human operator to investigate further. 
+
+### For future consideration:
+- The operator might need access to more broker metrics. We can easily implement additional endpoints to expose them.
+- The readiness check could also make use of the KafkaAgent endpoint to get the broker state instead of checking for a file on disk.
+- The liveness and readiness checks potentially will need to access non yammer metrics for KRAFT, which could be exposed in the same manner, but would need a new mechanism to collect them.
+
+## Affected/not affected projects
+
+* strimzi/kafka-cluster-operator
+
+## Compatibility
+
+When the operator gets updated with the KafkaRoller change, it will expect the KafkaAgent endpoint to be there. However, KafkaAgent would only get updated after the operator.
+
+1. Roll out the KafkaAgent changes for the new API.
+2. Roll out the KafkaRoller changes and gated behind a feature flag. By default, KafkaRoller will still have the current behaviour and will not use the KafkaAgent API. 
+3. User can enable KafkaRoller to use KafkaAgent API using the feature flag. The user has to make sure that they are running the version that has the KafkaAgent API before they enable the feature. 
+4. We can mark the feature flag deprecetad immediately. 
+5. In the next major release, we can remove the feature flag and KafkaRoller can always use the KafkaAgent API.
```

Comment:
I assume you mean Feature gate (as is the Strimzi terminology)? If you want to use a Feature Gate, you should suggest a name and a proposed schedule here. (check the other proposals with feature gates for examples). 

But I think you should try to figure out a better way because rolling out something like this will take a fairly long time. A very similar feature - Control plan listener was rolling out for almost 9 releases. It will be very complicated to build some further features on top of this in such case, maintain and test the code etc.

Reply:
It should be possible to use this opportunistically: If the endpoint works then apply the new logic, if not then apply the current logic.

Reply:
I guess that would be an option ... the thing I was wondering about is whether this is good enough with respect to distinguishing if the _agent is stuck_ or whether the broker uses the old version. 

I guess we also need to think about the `then apply the current logic`. The upgrade part is something hat will be needed for a fairly long time. So this needs to elaborate on how this will be handled in the code as well.

Reply:
I had the same concern @scholzj mentioned. However, I think if the agent is stuck, it's ok to apply the current logic that would force restart the broker, that way we might recover the stuck agent. 
I have addressed the suggestion in the latest commit.